readium/shared/src/main/java/org/readium/r2/shared/util/asset/AssetOpener.kt

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2023 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.shared.util.asset
+
+import java.io.File
+import org.readium.r2.shared.util.AbsoluteUrl
+import org.readium.r2.shared.util.Error
+import org.readium.r2.shared.util.Try
+import org.readium.r2.shared.util.Url
+import org.readium.r2.shared.util.format.Format
+import org.readium.r2.shared.util.format.FormatHints
+import org.readium.r2.shared.util.getOrElse
+import org.readium.r2.shared.util.mediatype.MediaType
+import org.readium.r2.shared.util.resource.Resource
+import org.readium.r2.shared.util.resource.ResourceFactory
+import org.readium.r2.shared.util.toUrl
+
+/**
+ * Retrieves an [Asset] instance providing reading access to the resource(s) of an asset stored at
+ * a given [Url] as well as a canonical media type.
+ */
+public class AssetOpener(
```

Comment:
How about merging `AssetOpener` and `AssetSniffer` and exposing only the `open()` API?

Reply:
It's tricky to handle errors this way. `SchemeNotSupported` doesn't make any sense with methods not taking a Url as input.